### PR TITLE
prov/efa: fix a bug in local read

### PR DESCRIPTION
**Pull request by wzamazon**
fix a bug in local read

### CODE REVIEW

File: prov/efa/src/rxr/rxr_pkt_cmd.c [hunk 1]
```diff
@@ -838,16 +838,15 @@ void rxr_pkt_handle_send_error(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entr
...
-		 * Either way, we need to ignore this error completion.
-		 */
+	/*
+	 * For a send completion, pkt_entry->addr can be FI_ADDR_NOTAVAIL in 3 situations:
+	 * 1. the pkt_entry is used for a local read operation
+	 * 2. a new peer with same gid+qpn was inserted to av, thus the peer was removed from AV.
+	 * 3. application removed the peer's address from av.
+	 * In 1, we should proceed. For 2 and 3, the send completion should be ignored.
+	 */
+	if (pkt_entry->addr == FI_ADDR_NOTAVAIL &&
```
Comment (PeterGottesman) [inline prov/efa/src/rxr/rxr_pkt_cmd.c:848]: Will this catch all of the failure modes described? I would think that you still want to have a call to `rxr_ep_get_peer` to verify that there is a peer with the address in `pkt_entry->addr`.

Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_cmd.c:848]: This function does not use `peer`. For all 3 scenario, `pkt_entry->addr`  will be FI_ADDR_NOTAVAIL

**Comment by PeterGottesman** [comment]
Quick and dirty grep shows there were one or two comments which referenced the two scenarios where peer may be null. Should those be updated?
```
% grep -B2 -A2 -nri "\* .*peer.*null" prov/efa/src/rxr
prov/efa/src/rxr/rxr_ep.c-2190-
prov/efa/src/rxr/rxr_ep.c-2191- /*
prov/efa/src/rxr/rxr_ep.c:2192:  * peer can be NULL when the pkt_entry is a RMA_CONTEXT_PKT,
prov/efa/src/rxr/rxr_ep.c-2193-  * and the RMA is a local read toward the endpoint itself
prov/efa/src/rxr/rxr_ep.c-2194-  */
--
prov/efa/src/rxr/rxr_ep.c-2244-
prov/efa/src/rxr/rxr_ep.c-2245- /*
prov/efa/src/rxr/rxr_ep.c:2246:  * peer can be NULL when:
prov/efa/src/rxr/rxr_ep.c-2247-  *
prov/efa/src/rxr/rxr_ep.c-2248-  * 1. the pkt_entry is a RMA_CONTEXT_PKT, and the RMA op is a local read
--
prov/efa/src/rxr/rxr_pkt_cmd.c-721-     if (!peer) {
prov/efa/src/rxr/rxr_pkt_cmd.c-722-             /*
prov/efa/src/rxr/rxr_pkt_cmd.c:723:              * If peer is NULL, it means the peer has been removed from AV.
prov/efa/src/rxr/rxr_pkt_cmd.c-724-              * In this case, ignore this error completion.
prov/efa/src/rxr/rxr_pkt_cmd.c-725-              */
--
prov/efa/src/rxr/rxr_pkt_cmd.c-837-     if (!peer) {
prov/efa/src/rxr/rxr_pkt_cmd.c-838-             /*
prov/efa/src/rxr/rxr_pkt_cmd.c:839:              * peer could be NULL in the following 2 scenarios:
prov/efa/src/rxr/rxr_pkt_cmd.c-840-              * 1. a new peer with same gid+qpn was inserted to av, thus the peer was remove.
prov/efa/src/rxr/rxr_pkt_cmd.c-841-              * 2. application removed the peer's address from av.
```
The above was run a few commits behind main so there may be false positives/wrong line numbers, but you get the idea.

**Comment by PeterGottesman** [comment]
Testing shows the original bug is resolved.

**Review by zhngaj — COMMENTED** [review]
The commit message says `... to indicate a pkt_entry is used by local read, and will not
send completion for such a pkt_entry.`. Why is it `will not send completion` ? I thought the send completion will NOT be ignored in this case.

**Comment by wzamazon** [comment]
@zhngaj Thanks! You are right, and the commit message is wrong. Fixed in new revision.

**Comment by wzamazon** [comment]
@PeterGottesman  The comment regarding peer being NULL is still valid. When pkt_entry->addr is `FI_ADDR_NOTAVAIL`, peer will be NULL.

**Comment by PeterGottesman** [comment]
> @PeterGottesman The comment regarding peer being NULL is still valid. When pkt_entry->addr is `FI_ADDR_NOTAVAIL`, peer will be NULL.

I just went through and looked again, it looks like it is just here that should be updated?
https://github.com/ofiwg/libfabric/blob/2eb975c59323e75395204d3ee9711f6ca4d748e0/prov/efa/src/rxr/rxr_ep.c#L2347-L2359

**Review by PeterGottesman — APPROVED** [review]

**Review by zhngaj — APPROVED** [review]

**Review by PeterGottesman — APPROVED** [review]

**Comment by wzamazon** [comment]
bot:aws:retest

**Pull request merged**